OSSpecific/src/artofillusion/osspecific/MacOSPlugin.java

```diff
@@ -40,15 +40,35 @@ public void processMessage(int message, Object args[])
       UIUtilities.setStandardDialogInsets(3);
       try
       {
-        // Use reflection to set up the application menu.
-
-        Class appClass = Class.forName("com.apple.eawt.Application");
-        Object app = appClass.getMethod("getApplication").invoke(null);
-        appClass.getMethod("setEnabledAboutMenu", Boolean.TYPE).invoke(app, Boolean.TRUE);
-        appClass.getMethod("setEnabledPreferencesMenu", Boolean.TYPE).invoke(app, Boolean.TRUE);
-        Class listenerClass = Class.forName("com.apple.eawt.ApplicationListener");
-        Object proxy = Proxy.newProxyInstance(appClass.getClassLoader(), new Class [] {listenerClass}, this);
-        appClass.getMethod("addApplicationListener", listenerClass).invoke(app, proxy);
+        if (System.getProperty("java.version").startsWith("1.8."))
+        {
+          // Use the old Apple specific API.
+
+          appleApi = true;
+          Class appClass = Class.forName("com.apple.eawt.Application");
+          Object app = appClass.getMethod("getApplication").invoke(null);
+          appClass.getMethod("setEnabledAboutMenu", Boolean.TYPE).invoke(app, Boolean.TRUE);
+          appClass.getMethod("setEnabledPreferencesMenu", Boolean.TYPE).invoke(app, Boolean.TRUE);
+          Class listenerClass = Class.forName("com.apple.eawt.ApplicationListener");
+          Object proxy = Proxy.newProxyInstance(appClass.getClassLoader(), new Class[]{listenerClass}, this);
+          appClass.getMethod("addApplicationListener", listenerClass).invoke(app, proxy);
+        }
+        else
+        {
+          // Use the Desktop API introduced in Java 9.
+
+          appleApi = false;
+          Class aboutHandlerClass = Class.forName("java.awt.desktop.AboutHandler");
+          Class openHandlerClass = Class.forName("java.awt.desktop.OpenFilesHandler");
+          Class preferencesHandlerClass = Class.forName("java.awt.desktop.PreferencesHandler");
+          Class quiteHandlerClass = Class.forName("java.awt.desktop.QuitHandler");
+          Object proxy = Proxy.newProxyInstance(Desktop.class.getClassLoader(), new Class[]{aboutHandlerClass, openHandlerClass, preferencesHandlerClass, quiteHandlerClass}, this);
+          Desktop desktop = Desktop.getDesktop();
```

Comment:
> Desktop desktop = Desktop.getDesktop();

How to expect compile this with JDK8? If compile this with 9 what the reason set handlers via proxy?

Reply:
The `Desktop` class itself has been around since Java 6, so this will compile. For the time being, or official builds will be made with Java 8, but we can't guarantee that it will be run with 8.